File: python/백준/그래프/BFS/2206 벽 부수고 이동하기.py
import sys
from collections import deque
input = sys.stdin.readline

N, M = map(int,input().split())

grid = [list(map(int, input().strip())) for _ in range(N)]
directions = [(1, 0), (0, 1), (-1, 0), (0, -1)]

visited = [[[0]*2] for _ in range(M) for __ in range(N)]
def bfs():

    queue = deque([0,0,0])
    visited[0][0][0] = 1

    while queue:
        x, y, broken = queue.popleft()

        if x == N-1 and y == M-1:
            return visited[x][y][broken]
        
        for dx, dy in directions:
            nx, ny = dx + x, dy + y

        if 0 <= nx < N and 0 <= ny < M:
            if grid[nx][ny] == 0 and visited[nx][ny][broken] == 0:
                visited[nx][ny][broken] = visited[x][y][broken] + 1
                queue.append((x, y, broken))
            elif grid[nx][ny] == 1 and visited[nx][ny][1] == 0 and broken == 0:
                visited[nx][ny][1] = visited[x][y][broken] + 1
                queue.append((x, y, 1))

ans = bfs()
print(ans)

File: cpp/백준/그래프/BFS/2206 벽 부수고 이동하기.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <string>
#include <tuple>
using namespace std;

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(nullptr);

    int N, M;
    cin >> N >> M;

    vector<vector<int>> grid(N, vector<int>(M));
    for (int i = 0; i < N; ++i) {
        string row;
        cin >> row;
        for (int j = 0; j < M; ++j) grid[i][j] = row[j] - '0';
    }

    int dx[] = {1, 0, -1, 0};
    int dy[] = {0, 1, 0, -1};

    vector<vector<vector<int>>> visited(N, vector<vector<int>>(M, vector<int>(2, 0)));
    queue<tuple<int, int, int>> q;
    q.push({0, 0, 0});
    visited[0][0][0] = 1;

    int ans = -1;
    while (!q.empty()) {
        auto [x, y, broken] = q.front();
        q.pop();

        if (x == N - 1 && y == M - 1) {
            ans = visited[x][y][broken];
            break;
        }

        for (int i = 0; i < 4; ++i) {
            int nx = x + dx[i];
            int ny = y + dy[i];
            if (nx < 0 || nx >= N || ny < 0 || ny >= M) continue;
            if (grid[nx][ny] == 0 && visited[nx][ny][broken] == 0) {
                visited[nx][ny][broken] = visited[x][y][broken] + 1;
                q.push({nx, ny, broken});
            } else if (grid[nx][ny] == 1 && broken == 0 && visited[nx][ny][1] == 0) {
                visited[nx][ny][1] = visited[x][y][broken] + 1;
                q.push({nx, ny, 1});
            }
        }
    }

    cout << ans << "\n";
    return 0;
}
